controllers/lvmcluster_controller.go

```diff
@@ -292,37 +334,32 @@ func (r *LVMClusterReconciler) updateLVMClusterStatus(ctx context.Context, insta
 	return nil
 }
 
-func (r *LVMClusterReconciler) getExpectedVgCount(ctx context.Context, instance *lvmv1alpha1.LVMCluster) (int, error) {
-
-	var vgCount int
-
-	nodeList := &corev1.NodeList{}
-	err := r.Client.List(ctx, nodeList)
-	if err != nil {
-		r.Log.Error(err, "failed to list Nodes")
-		return 0, err
-	}
+func (r *LVMClusterReconciler) nodesByDeviceClass(allNodes *corev1.NodeList, instance *lvmv1alpha1.LVMCluster) (map[string][]corev1.Node, error) {
+	nodesForDeviceClasses := map[string][]corev1.Node{}
 
 	for _, deviceClass := range instance.Spec.Storage.DeviceClasses {
 		if deviceClass.NodeSelector == nil {
-			vgCount += len(nodeList.Items)
+			nodesForDeviceClasses[deviceClass.Name] = allNodes.Items
 			continue
 		}
 
-		for i := range nodeList.Items {
-			matches, err := corev1helper.MatchNodeSelectorTerms(&nodeList.Items[i], deviceClass.NodeSelector)
+		nodes := make([]corev1.Node, len(instance.Spec.Storage.DeviceClasses))
+		for i := range allNodes.Items {
+			matches, err := corev1helper.MatchNodeSelectorTerms(&allNodes.Items[i], deviceClass.NodeSelector)
 			if err != nil {
-				r.Log.Error(err, "failed to match node selector")
-				return 0, err
```

Comment:
this is to allow other deviceClasses to succeed even though there might be one deviceClass with a failing Selector. Otherwise all deviceClasses will start to loose their status once one selector is faulty. I wanted to avoid this scenario by not failing out here but skipping that deviceClass and continuing for others.

Reply:
the log message currently is the only reliable way. I think the status of LVMCluster is not reliable enough here since it doesnt check degraded status correctly.

Reply:
So I have one primary question: since we are now just continuing past failing deviceClasses, does this essentially just become a warning? I'm wondering if we still need to error out in some way here so that the user knows they have a malformed or misconfigured deviceClass

Reply:
So what happens is that the status will not contain this deviceClass because there is no node found for it. This will cause the status to not pick it up at all. Looking at what I just wrote, I will reintroduce the error condition here since that seems like a bad idea :D

Reply:
I decided to rollback my controller refactoring of the status to make this PR leaner and check this in a separate issue